btp-document/Smart-Contracts-ICON.md

```diff
@@ -0,0 +1,387 @@
+  - `mtaIsAllowNewerWitness`: allow to verify newer witness. This setting allows BMV to verify in case of MTA block height of a client is higher than MTA block height of BMV contract
+  - `relayLastBlockHash`: hash of previous block - `relayMtaOffset`. BMV must check that a previous hash of an incoming block is equal to the `relayLastBlockHash`
+  - `paraLastBlockHash`: hash of previous block - `paraMtaOffset`. BMV must check that a previous hash of an incoming block is equal to the `paraLastBlockHash`
+  - `encodedValidators`: `Base64(RLP.encode(List<byte[]> validatorPublicKey))`, encoded of validators list of relay chain
+  - `relayEventDecoderAddress`: address of Event Decoder for Relaychain - e.g. ***Kusama*** Event Decoder
+  - `paraEventDecoderAddress`: address of Event Decoder for Parachain - e.g. ***Moonriver*** Event Decoder
+  - `relayCurrentSetId`: set a current counter of updating `encodedValidators`. When a list of Validators is updated, the setID is increased by one
+  - `paraChainId`: an ID of Parachain
+
+For a sake of simplicity, we have setup an utility that helps to initialize these parameters. Please follow the instructions below:
+
+- Install Yarn packages:
+
+```shell
+$ cd $CONFIG_DIR/btp/javascore/bmv/helper
+$ yarn
+```
+
+- Specify your configurations:
+
+  - Open a file - `getBMVInitializeParams.ts`
+
+  - Specify Relaychain and Parachain web socket endpoint by replacing `RELAY_ENDPOINT`, `PARA_ENDPOINT` in line 30, and 31 with your configuration. In this example, we use a local testnet, thus the settings are `PARA_ENDPOINT = ws://127.0.0.1:9944`, and `RELAY_ENDPOINT = wss://kusama-rpc.polkadot.io`
+
+  - In order to setting `relayMtaOffset`, and `paraMtaOffset`, you replace `RELAY_OFFSET`, `PARA_OFFSET` in line 33, 34 with the current block height of corresponding chains. You can use [Polkadot Portal](https://polkadot.js.org/apps/?#/explorer) (using Chrome browser) to retrieve these information. Since we are using a local testnet for this example, `paraMtaOffset` is simply set by checking a current block of deployed node and `relayMtaOffset` is the current block of ***Kusama*** that is retrieved via the above link
+
+- Then, run the below command:
+
+```shell
+$ yarn getBMVInitializeParams
+```
+
+- In success, a file, `BMVInitializeData.json`, will be generated in the `$CONFIG_DIR/javascore/bmv/helper` directory. This JSON file contains initialized parameters that are essential to deploy ***BMV*** contract. Please copy and paste them to the below command:
+
+```shell
+source rpc.sh
+rpcch icon
+rpcks $GOLOOP_KEY_STORE $GOLOOP_KEY_SECRET
+goloop rpc sendtx deploy ./javascore/parachain-BMV-optimized.jar \
+    --content_type application/java --step_limit 13610920001 \
+    --param relayMtaOffset=${Relaychain Offset} \
+    --param paraMtaOffset=${Parachain Offset} \
+    --param bmc=$(cat bmc.$(rpcch)) \
+    --param net=$(cat net.btp.icon) \
+    --param mtaRootSize=${MTA root size} \
```

Comment:
Need to explain which value to set for `mtaRootSize`, `mtaCacheSize`, `mtaIsAllowNewerWitness`

Reply:
Agree